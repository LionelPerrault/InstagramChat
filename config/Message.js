// HANDY METHODS FOR MESSAGE ROUTES

const db = require('./db'),
  { promisify } = require('util'),
  { unlink } = require('fs'),
  root = process.cwd()

/**
 * Returns last message time of a comversation
 * @param {Number} con_id Conversation ID
 */
const getLastMssgTime = async con_id => {
  let s = await db.query(
    'SELECT MAX(message_time) AS ti FROM messages WHERE con_id = ?',
    [con_id]
  )
  return s[0].ti
}

/**
 * Returns last message of a comversation
 * @param {Number} con_id Conversation ID
 */
const getLastMssg = async con_id => {
  let [{ last }] = await db.query(
    'SELECT MAX(message_id) AS last FROM messages WHERE con_id = ?',
    [con_id]
  )
  let l = await db.query(
    'SELECT message, type, mssg_by FROM messages WHERE message_id=?',
    [last]
  )
  return l[0]
}

/**
 * Deletes a conversation
 * @param {Number} con_id Conversation ID
 */
const deleteCon = async con_id => {
  try {
    let messages = await db.query(
        'SELECT message, type FROM messages WHERE con_id=?',
        [con_id]
      ),
      deleteMessageFile = promisify(unlink)

    for (let m of messages) {
      if (m.type != 'text') {
        await deleteMessageFile(`${root}/dist/messages/${m.message}`)
      }
    }

    await db.query('DELETE FROM messages WHERE con_id=?', [con_id])
    await db.query('DELETE FROM conversations WHERE con_id=?', [con_id])
  } catch (error) {
    console.log(error)
  }
}

module.exports = {
  getLastMssgTime,
  getLastMssg,
  deleteCon,
}
