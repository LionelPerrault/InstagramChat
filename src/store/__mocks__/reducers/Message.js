export default {
  onlineUsers: [
    {
      user: 16,
      username: 'zayn',
      firstname: 'Zayn',
      surname: 'Malik',
      mutualUsersCount: 0,
    },
    {
      user: 7,
      username: 'ghalib',
      firstname: 'Mirza',
      surname: 'Ghalib',
      mutualUsersCount: 1,
    },
  ],
  conversations: [
    {
      con_id: 24,
      con_with: 7,
      con_with_username: 'ghalib',
      con_with_firstname: 'Mirza',
      con_with_surname: 'Ghalib',
      unreadMssgs: 0,
      lastMssg: {
        lastMssgTime: '1480114098767',
        lastMessage: 'eo',
        lastMssgType: 'text',
        lastMssgBy: 7,
      },
    },
    {
      con_id: 40,
      con_with: 10,
      con_with_username: 'noddy',
      con_with_firstname: 'your',
      con_with_surname: 'noddy',
      unreadMssgs: 0,
      lastMssg: {
        lastMssgTime: null,
        lastMessage: '',
        lastMssgType: '',
        lastMssgBy: '',
      },
    },
  ],
  conDetails: {
    con_id: 29,
    con_with: 28,
    con_with_username: 'selena',
    con_with_firstname: 'selena',
    con_with_surname: 'gomez',
    isOnline: false,
    lastOnline: '1480114098767',
  },
  conAbout: {
    mssgsCount: 6,
    media: [
      {
        imgSrc: 'instagram_message_1528278864884.jpg',
        mssg_by: 24,
        mssg_by_username: 'takkar',
      },
    ],
    con_time: '1480114098767',
    mutualFollowersCount: 1,
  },
  messages: [
    {
      message_id: 93,
      con_id: 29,
      mssg_by: 28,
      mssg_to: 24,
      message: 'instagram_message_1525091176544.jpg',
      type: 'sticker',
      status: 'read',
      message_time: '1480114098767',
    },
    {
      message_id: 94,
      con_id: 29,
      mssg_by: 24,
      mssg_to: 28,
      message: 'kjkjk',
      type: 'text',
      status: 'unread',
      message_time: '1480114098767',
    },
    {
      message_id: 99,
      con_id: 29,
      mssg_by: 24,
      mssg_to: 28,
      message: 'instagram_message_1525809886884.jpg',
      type: 'sticker',
      status: 'unread',
      message_time: '1480114098767',
    },
    {
      message_id: 102,
      con_id: 29,
      mssg_by: 24,
      mssg_to: 28,
      message: 'instagram_message_1528278864884.jpg',
      type: 'image',
      status: 'unread',
      message_time: '1480114098767',
    },
  ],
  unreadMessages: 0,
}
