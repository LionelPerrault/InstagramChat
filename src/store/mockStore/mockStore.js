import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import logger from 'redux-logger'
import promise from 'redux-promise-middleware'

// mock-reducers
import User from './mock-reducers/User'
import Notification from './mock-reducers/Notification'
import Group from './mock-reducers/Group'
import Follow from './mock-reducers/Follow'
import Explore from './mock-reducers/Explore'

const initialState = {
  User,
  Notification,
  Group,
  Follow,
  Explore
}

const middlewares = [
  promise(), thunk, logger
]
const createStore = configureStore(middlewares)
const mockStore = createStore(initialState)

export default mockStore
