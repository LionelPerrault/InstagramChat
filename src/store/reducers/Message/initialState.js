export default {
  conversations: [],
  messages: [],
  conAbout: {},
  unreadMessages: 0,
  onlineUsers: [],
  conDetails: {
    con_with: null,
    con_with_username: '',
    con_with_firstname: '',
    con_with_surname: '',
    isOnline: 'no',
    lastOnline: '',
  },
}
