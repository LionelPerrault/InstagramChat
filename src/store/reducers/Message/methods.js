export const addCon = (cons, con) => {
  cons = [con, ...cons]
  return cons
}

export const message = (messages, message) => {
  messages = [...messages, message]
  return messages
}

export const changeLastMssg = (conversations, mssgDet) => {
  return conversations.map(c => {
    if (c.con_id == mssgDet.con_id) {
      c.lastMssg = mssgDet.lastMssg
    }
    return c
  })
}

export const deleteMssg = (messages, message_id) =>
  messages.filter(m => m.message_id != message_id)

export const unsendAllMssgs = (messages, mssg_by) =>
  messages.filter(m => m.mssg_by != mssg_by)

export const deleteConversation = (cons, con_id) =>
  cons.filter(c => c.con_id != con_id)

export const readCon = (cons, con_id) => {
  return cons.map(c => {
    if (c.con_id == con_id) {
      c.unreadMssgs = 0
    }
    return c
  })
}
