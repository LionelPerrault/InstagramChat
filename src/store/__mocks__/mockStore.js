import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

// mock-reducers
import User from './reducers/User'
import Notification from './reducers/Notification'
import Group from './reducers/Group'
import Follow from './reducers/Follow'
import Explore from './reducers/Explore'
import Post from './reducers/Post'
import Hashtag from './reducers/Hashtag'
import Setting from './reducers/Setting'
import Message from './reducers/Message'

const initialState = {
  User,
  Notification,
  Group,
  Follow,
  Explore,
  Post,
  Hashtag,
  Setting,
  Message,
}

const middlewares = [thunk]
const createStore = configureStore(middlewares)
const mockStore = createStore(initialState)

export default mockStore
