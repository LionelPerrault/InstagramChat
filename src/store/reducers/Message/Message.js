/* eslint indent:0 */
/* eslint no-unreachable:0 */

import initialState from './initialState'
import * as methods from './methods'

export default (state = initialState, action) => {
  let py = action.payload

  switch (action.type) {
    case 'GET_CONVERSATIONS':
      return { ...state, conversations: py }
      break

    case 'CONVERSATION_ADDED':
      return {
        ...state,
        conversations: methods.addCon(state.conversations, py),
      }
      break

    case 'GET_CONVERSATION_MESSAGES':
      return { ...state, messages: py }
      break

    case 'MESSAGE':
      return {
        ...state,
        messages: methods.message(state.messages, py),
      }
      break

    case 'CHANGE_LAST_MSSG':
      return {
        ...state,
        conversations: methods.changeLastMssg(state.conversations, py, 'user'),
      }
      break

    case 'DELETE_MESSAGE':
      return {
        ...state,
        messages: methods.deleteMssg(state.messages, py),
      }
      break

    case 'UNSEND_ALL_MSSGS':
      return {
        ...state,
        messages: methods.unsendAllMssgs(state.messages, py),
      }
      break

    case 'DELETE_CONVERSATION':
      return {
        ...state,
        conversations: methods.deleteConversation(state.conversations, py),
      }
      break

    case 'GET_CONVERSATION_ABOUT':
      return { ...state, conAbout: py }
      break

    case 'GET_UNREAD_MESSAGES':
      return { ...state, unreadMessages: py }
      break

    case 'READ_CONVERSATION':
      return {
        ...state,
        conversations: methods.readCon(state.conversations, py),
      }
      break

    case 'UPDATE_UNREAD_CONVERSATIONS':
      return {
        ...state,
        unreadMessages: state.unreadMessages - py,
      }
      break

    case 'GET_ONLINE_USERS':
      return { ...state, onlineUsers: py }
      break

    case 'GET_CON_DETAILS':
      return { ...state, conDetails: py }
      break
  }
  return state
}
