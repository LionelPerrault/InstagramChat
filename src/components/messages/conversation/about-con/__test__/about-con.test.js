import React from 'react'
import { create } from 'react-test-renderer'
import { Provider } from 'react-redux'
import mockStore from '../../../../../store/__mocks__/mockStore'
import AboutConversation, { PureAboutConversation } from '../about-con'
import { shallow } from 'enzyme'
import Message from '../../../../../store/__mocks__/reducers/Message'

describe('AboutConversation Component', () => {
  let mockFn = jest.fn()

  it('should match snapshot', () => {
    const tree = create(
      <Provider store={mockStore}>
        <AboutConversation back={mockFn} />
      </Provider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should hide spinner when loading = false', () => {
    const wrapper = shallow(
      <PureAboutConversation
        back={mockFn}
        cd={Message.conDetails}
        dispatch={mockFn}
      />
    )
    wrapper.setState({ loading: false })
    expect(wrapper.find('IsLoading').prop('loading')).toBe(false)
  })
})
