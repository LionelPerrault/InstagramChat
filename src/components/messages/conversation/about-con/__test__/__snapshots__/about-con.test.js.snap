// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutConversation Component should match snapshot 1`] = `
Array [
  <div
    className="overlay"
    onClick={null}
    style={
      Object {
        "cursor": "inherit",
        "opacity": 0.5,
      }
    }
  />,
  <div
    className="modal modal_big"
  >
    <div
      style={
        Object {
          "animation": "bcCCNc 300ms ease 0s 1 normal none running",
          "backfaceVisibility": "visible",
        }
      }
    >
      <div
        className="modal_header"
      >
        <span
          className="title"
        >
          About conversation
        </span>
      </div>
      <div
        className="modal_middle"
        style={
          Object {
            "height": 450,
            "overflow": "hidden",
            "position": "relative",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "bottom": 0,
              "left": 0,
              "marginBottom": 0,
              "marginRight": 0,
              "overflow": "scroll",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <div
            className="spinner"
          >
            <span />
            <span />
            <span />
          </div>
          <div
            className="modal_main cLoading"
            style={
              Object {
                "padding": 0,
              }
            }
          >
            <div
              className="about_con"
            >
              <div
                className="sli_with_div"
              >
                <span
                  className="sli_label"
                >
                  Conversation with
                </span>
                <div
                  className="sli_with"
                >
                  <img
                    src="/users/28/avatar.jpg"
                  />
                  <div
                    className="sli_with_cont"
                  >
                    <Link
                      to="/profile/selena"
                    >
                      selena
                    </Link>
                    <span
                      className="sli_w"
                    >
                      1 mutual follower
                    </span>
                  </div>
                </div>
              </div>
              <div
                className="sli_time"
              >
                <span
                  className="sli_label"
                >
                  Conversation since
                </span>
                <span
                  className="sli_bold"
                >
                  a month ago
                </span>
              </div>
              <div
                className="sli_mssg_count"
              >
                <span
                  className="sli_label"
                >
                  No. of messages
                </span>
                <span
                  className="sli_bold"
                >
                  6 messages
                </span>
              </div>
              <div
                className="sli_media"
              >
                <span
                  className="sli_label"
                >
                  Media
                </span>
                <img
                  className="sli_media_img"
                  src="/messages/instagram_message_1528278864884.jpg"
                  title="By takkar"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "borderRadius": 3,
              "bottom": 2,
              "display": "none",
              "height": 6,
              "left": 2,
              "position": "absolute",
              "right": 2,
            }
          }
        >
          <div
            style={
              Object {
                "backgroundColor": "rgba(0,0,0,.2)",
                "borderRadius": "inherit",
                "cursor": "pointer",
                "display": "block",
                "height": "100%",
                "position": "relative",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "borderRadius": 3,
              "bottom": 2,
              "display": "none",
              "position": "absolute",
              "right": 2,
              "top": 2,
              "width": 6,
            }
          }
        >
          <div
            style={
              Object {
                "backgroundColor": "rgba(0,0,0,.2)",
                "borderRadius": "inherit",
                "cursor": "pointer",
                "display": "block",
                "position": "relative",
                "width": "100%",
              }
            }
          />
        </div>
      </div>
      <div
        className="modal_bottom"
      >
        <a
          className="pri_btn"
          href="#"
          onClick={[Function]}
        >
          Back
        </a>
      </div>
    </div>
  </div>,
]
`;
